### PR TITLE
SSAO demo adopts EffectComposer and resize support

### PR DESCRIPTION
**Pull request by daoshengmu**
Using `EffectComposer` to refactor this postproessing demo, and fixing memory issue from `EffectComposer`. Currently, when using `EffectComposer`, we still need to include `CopyShader.js` and `MaskPass.js`. I think CopyShader is required, it can help us copy pixel from render target. But, `MaskPass.js` is because `webgl_postprocessing_advanced` demo has to access copyShader's stencil function. I think we can consider If we should keep it or not. @WestLangley Please take a look at.

### CODE REVIEW

**Comment by WestLangley** [comment]
It appears there are bugs. If I resize the window when postprocessing is enabled, the screen turns black.

> But, MaskPass.js is because webgl_postprocessing_advanced demo has to access copyShader's stencil function

`MaskPass.js` is required because `EffectComposer` references `THREE.MaskPass`.

**Comment by daoshengmu** [comment]
Regarding to ```THREE.MaskPass```, I have searched the whole project. That is only used by ```webgl_postprocessing_advanced``` demo. It occurs the side effect is every demo has to include ```MaskPass.js``` while they want to use ```EffectComposer```. This is just my personal question, I don't really force to remove it.

@WestLangley . I have checked it on Firefox Nightly and Chrome, but I didn't find the resize issue that you mentioned.

**Comment by WestLangley** [comment]
The following line in `onWindowResize()` results in a blank screen for me on OS X 10.10.3 Chrome 43.0.2357.132 and Safari 8.0.6.

    effectComposer.setSize( window.innerWidth, window.innerHeight );

/ping @benaadams Do you mind having a look at the changes to `EffectComposer` in this PR?

>  It occurs the side effect is every demo has to include MaskPass.js while they want to use EffectComposer. 

That is because of this line in `EffectComposer.js`:

    if ( pass instanceof THREE.MaskPass ) {

**Comment by benaadams** [comment]
Will have a look; general rendertarget resize was fixed by #6355 including dispose so a lot of the extra ```EffectComposer``` is no longer necessary which has been picked up in this PR.

However, with the defunct ```reset``` no longer being called might be lacking the pixel ratio for hpi on retina? Will add code comment.

File: examples/js/postprocessing/EffectComposer.js [hunk 1]
```diff
@@ -127,12 +132,8 @@ THREE.EffectComposer.prototype = {
 
 	setSize: function ( width, height ) {
```
Comment (benaadams) [inline examples/js/postprocessing/EffectComposer.js]: Params need to be adjusted by  renderer.getPixelRatio()?

Comment (benaadams) [inline examples/js/postprocessing/EffectComposer.js]: As here: https://github.com/mrdoob/three.js/pull/6820/files#diff-f69be6f1b9bed3c3d7387ed929cd29faL13
But not done on resize here: https://github.com/mrdoob/three.js/pull/6820/files#diff-76f301411c5013d33c69fa37efb52263R146

Comment (benaadams) [inline examples/js/postprocessing/EffectComposer.js]: Previously done here: https://github.com/mrdoob/three.js/pull/6820/files#diff-f69be6f1b9bed3c3d7387ed929cd29faR113 however this function is now defunct.

**Comment by daoshengmu** [comment]
I also hesitate that we should adjust size by pixel ratio in this ```setSize()``` function or we can suppose it has been adjusted before set to be the params. Currently, I think that adjusting pixel ratio before sending to ```setSize()``` might be a better way. Because we also have other render targets need to be adjusted like this: 
```
var pixelRatio = renderer.getPixelRatio();
var newWidth  = Math.floor( window.innerWidth / pixelRatio ) || 1;
var newHeight = Math.floor( window.innerHeight / pixelRatio ) || 1;
depthRenderTarget.setSize( newWidth, newHeight );
effectComposer.setSize( newWidth, newHeight );
```

Moreover, I don't understand what you mean about "Previously done before this PR here: https://github.com/mrdoob/three.js/pull/6820/files#diff-f69be6f1b9bed3c3d7387ed929cd29faR113 however this function is now defunct." Actually, It works well and still refers pixel ratio even though I have  modified these lines.

**Comment by daoshengmu** [comment]
@WestLangley . I can reproduce this problem only on my Safari but Chrome works well. My Chrome version is the same with yours. Safari is v.8.0.7 and OS X is 10.10.4. I think the problem is because resizing window, it will dispose render targets and create new ones. When I stop resizing the window, the screen is normal and no memory issue. Therefore, in my opinion, it is not a big deal.

**Comment by WestLangley** [comment]
When I resize the window, the screen turns black and remains black.

Also, I do not understand your mods to `EffectComposer.reset()`. You are not calling that method. What is the purpose of the changes?

**Comment by benaadams** [comment]
> Moreover, I don't understand what you mean about...

```setSize``` used to call ```reset``` however it doesn't after this PR:

https://github.com/mrdoob/three.js/pull/6820/files#diff-f69be6f1b9bed3c3d7387ed929cd29faL135

Therefore the pixel ratio needs to be brought into ```setSize``` as it is done in the initalise; as initalise and resize need to match in use of pixel ratio.

**Comment by WestLangley** [comment]
The changes to `EffectComposer` in this PR appear to be causing problems in other examples when resizing.

screen shot 2015-07-11 at 9 16 20 am

Perhaps you should remove the changes to `EfffectComposer` from this PR and focus only on the SSAO demo. Then file a separate bug report regarding the `EffectComposer` memory issue.

**Comment by daoshengmu** [comment]
@WestLangley . I have got the point and fixed the resizing problem on ```webgl_line_colors``` demo. That is because it calls ```EffectComposer.reset()``` while it wants to resize the window.

Well, let me explain the reasons. That is all because ```THREE.WebGLRenderTarget``` has to be recreate the ```__webglFramebuffer``` while we want to resize it. You can use the clone function to ask it create a new one, but you MUST remember to dispose the older one, otherwise, it will happen memory problems. Or, you can call ```THREE.WebGLRenderTarget.setSize()```, it will help us resize and dispose the old one. I have solve these problems. Please expect my next commit.

@benaadams . I would like to confirm with you about the usage of ```EffectComposer```. As you mentioned, if we want users resize the EffectComposr, they should set the pixel ratio. As your original design, if we call ```THREE.EffectComposer.setSize()```, it will go to ```reset()```. However, It will NOT use renderer's pixel ratio because ```this.reset( renderTarget );``` will always send a valid renderTarget, and it will not be into the ```if ( renderTarget === undefined ) {``` condition. I have two options to enhance it. 

First, we can add these lines in ```EffectComposet.setSize()```

```
var pixelRatio = this.renderer.getPixelRatio();
var width  = Math.floor( this.renderer.context.canvas.width  / pixelRatio );
var height = Math.floor( this.renderer.context.canvas.height / pixelRatio );
this.renderTarget1.setSize( width, height );
this.renderTarget2.setSize( width, height );
```

Second, we can choose to disable ```EffectComposet.setSize()``` function for resize usage because ``````EffectComposet.reset()``` would cover this task. The good example is  ```webgl_line_colors``` demo.

**Comment by daoshengmu** [comment]
@WestLangley . I have followed your recommendation to remove the ```EffectCompser``` commit first. And I can reproduce the black effect while resize window by using my ```EffectComposer```, after I fetch the dev repo. I think this is due to some commits that were merged during we discuss my pull request. If you check my pull request is ok, we can ask mr.doob merge it first. And then, I will go to fix ```EffectComposer``` memory issue. Thanks

**Comment by WestLangley** [comment]
That sounds like a good idea.

**Comment by mrdoob** [comment]
Thanks guys!

**Pull request merged**